tsdb/engine/tsm1/encoding.go

```diff
@@ -298,36 +302,36 @@ func DecodeFloatBlock(block []byte, a []FloatValue) ([]FloatValue, error) {
 	tb, vb := unpackBlock(block)
 
 	// Setup our timestamp and value decoders
-	dec := NewTimeDecoder(tb)
-	iter, err := NewFloatDecoder(vb)
-	if err != nil {
+	tdec.Init(tb)
+	if err := vdec.SetBytes(vb); err != nil {
 		return nil, err
 	}
 
 	// Decode both a timestamp and value
 	i := 0
-	for dec.Next() && iter.Next() {
-		ts := dec.Read()
-		v := iter.Values()
-		if i < len(a) {
-			a[i].unixnano = ts.UnixNano()
-			a[i].value = v
+	for tdec.Next() && vdec.Next() {
+		ts := tdec.Read()
+		v := vdec.Values()
+		if i < len(*a) {
+			elem := &(*a)[i]
+			elem.unixnano = ts.UnixNano()
+			elem.value = v
 		} else {
-			a = append(a, FloatValue{ts.UnixNano(), v})
+			*a = append(*a, FloatValue{ts.UnixNano(), v})
```

Comment:
What would you think about adding a `Len()` method to the `TimeDecoder` interface, and then using that to ensure/allocate the necessary length of `a` here?

Currently, all times are decoded fully during the `decode()` in `tdec.Init(tb)`, so it would be as simple as `return len(d.ts)`

Additionally, you could eliminate the `v` field in `decoder` and shuffle the logic in `Next()` so that `Read()` just uses the current index, rather than re-slicing. And since we just convert the timestamp to a `time.Time` only to call `UnixNano` on it, maybe add a `ReadNano()` to the `TimeDecoder` interface as well? I'm not sure any of that would have a significant effect on allocations, other than maybe knowing the length ahead of time.